Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -978,143 +963,202 @@ extension SwiftLanguageServer {
+          self.rangeLimit = rangeLimit
+          self.lineFoldingOnly = lineFoldingOnly
+          super.init(viewMode: .sourceAccurate)
         }
 
-        guard let dict = result.success else {
-          req.reply(.failure(ResponseError(result.failure!)))
-          return
+        override func visit(_ node: SourceFileSyntax) -> SyntaxVisitorContinueKind {
+          return .visitChildren
         }
 
-        guard let syntaxMap: SKDResponseArray = dict[self.keys.syntaxmap],
-              let substructure: SKDResponseArray = dict[self.keys.substructure] else {
-          return req.reply([])
+        override func visit(_ node: TokenSyntax) -> SyntaxVisitorContinueKind {
+          // Index comments, so we need to see at least '/*', or '//'.
+          if node.leadingTriviaLength.utf8Length > 2 {
+            self.addTrivia(from: node, node.leadingTrivia)
+          }
+
+          if node.trailingTriviaLength.utf8Length > 2 {
+            self.addTrivia(from: node, node.trailingTrivia)
+          }
+
+          return .visitChildren
         }
 
-        /// Some ranges might occur multiple times.
-        /// E.g. for `print("hi")`, `"hi"` is both the range of all call arguments and the range the first argument in the call.
-        /// It doesn't make sense to report them multiple times, so use a `Set` here.
-        var ranges: Set<FoldingRange> = []
+        private func addTrivia(from node: TokenSyntax, _ trivia: Trivia) {
+          var start = node.position.utf8Offset
+          var lineCommentStart: Int? = nil
+          func flushLineComment(_ offset: Int = 0) {
+            if let lineCommentStart = lineCommentStart {
+              _ = self.addFoldingRange(
+                start: lineCommentStart,
+                end: start + offset,
+                kind: .comment)
+            }
+            lineCommentStart = nil
+          }
 
-        var hasReachedLimit: Bool {
-          let capabilities = self.clientCapabilities.textDocument?.foldingRange
-          guard let rangeLimit = capabilities?.rangeLimit else {
-            return false
+          for piece in node.leadingTrivia {
+            defer { start += piece.sourceLength.utf8Length }
+            switch piece {
+            case .blockComment(_):
+              flushLineComment()
+              _ = self.addFoldingRange(
+                start: start,
+                end: start + piece.sourceLength.utf8Length,
+                kind: .comment)
+            case .docBlockComment(_):
+              flushLineComment()
+              _ = self.addFoldingRange(
+                start: start,
+                end: start + piece.sourceLength.utf8Length,
+                kind: .comment)
+            case .lineComment(_), .docLineComment(_):
+              if lineCommentStart == nil {
+                lineCommentStart = start
+              }
+            case .newlines(1), .carriageReturns(1), .spaces(_), .tabs(_):
+              if lineCommentStart != nil {
```

Comment:
IIUC this means that line comments separated by multiple newlines will be considered as one folding range. E.g. the following will be considered as one folding range instead of three, right? (I think these should be three separate folding ranges).

```swift
// One line
// Followed by a line

// Another line
// Followed by another line


// An another line
func foo() {}
```

---

Also, does this mean that we are reporting folding ranges for single line comments like the following? (I think we should not).

```swift
// Single line comment
func foo() {}
```

Reply:
There are two newlines here

```
// One line
// Followed by a line

// Another line
// Followed by another line
```

and three here

```
// Another line
// Followed by another line


// An another line
```

These comments are therefore not merged.

Reply:
> Also, does this mean that we are reporting folding ranges for single line comments like the following? 

Outside of line range mode, both SourceKit and this PR report that single-line comment as a folding range.

Reply:
Do we have a test case that we are not merging these comments into one folding range?

Reply:
Yes, these two lines assert what you're looking for https://github.com/apple/sourcekit-lsp/blob/main/Tests/SourceKitLSPTests/FoldingRangeTests.swift#L73-L74

The corresponding comments are

```
//
// MARK: - A mark! -
//

//
// FIXME: a fixme
//
```

Reply:
Nice. Thank you 👍